**QUESTION:**

Grayscale images of 15 subjects under six different conditions were obtained in Yale university (known as yalefaces data set) and is given in the file yalefaces.zip. Due to storage limitations, only one representative image can be stored for each subject in the database for future automated facial recognition purpose. PCA is used to obtain the representative image for each person. (For this purpose, each image is converted to a column vector of pixel intensities by stacking each column of the image intensities one below the other and PCA is applied to the resulting data matrix of N x p where N is total number of pixels and p is number of images for each person). Given any image, the facial recognition method is based on the smallest Euclidean distance between the image and the representative images in the database. Determine the number of images out of 90 that you are able to correctly identify based on this approach. Use python opencv to read the images and reshape function to convert a matrix into a vector or vice versa.

In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/removed-yale/yalefaces/s03norm.png
/kaggle/input/removed-yale/yalefaces/s06ng.png
/kaggle/input/removed-yale/yalefaces/s06cl.png
/kaggle/input/removed-yale/yalefaces/s13glass.png
/kaggle/input/removed-yale/yalefaces/s02cl.png
/kaggle/input/removed-yale/yalefaces/s02norm.png
/kaggle/input/removed-yale/yalefaces/s04cl.png
/kaggle/input/removed-yale/yalefaces/s11ng.png
/kaggle/input/removed-yale/yalefaces/s04ng.png
/kaggle/input/removed-yale/yalefaces/s10ng.png
/kaggle/input/removed-yale/yalefaces/s07ng.png
/kaggle/input/removed-yale/yalefaces/s13ng.png
/kaggle/input/removed-yale/yalefaces/s14cl.png
/kaggle/input/removed-yale/yalefaces/s05cl.png
/kaggle/input/removed-yale/yalefaces/s04glass.png
/kaggle/input/removed-yale/yalefaces/s01cl.png
/kaggle/input/removed-yale/yalefaces/s08norm.png
/kaggle/input/removed-yale/yalefaces/s08happy.png
/kaggle/input/removed-yale/yalefaces/s05happy.png
/kaggle/input/removed-yale/yalefaces/s03ll.png
/kaggle/input/removed-yale/yalefaces/s15gl

In [3]:
# Importing necessary libraries

import cv2
import numpy as np
from sklearn.decomposition import PCA
from PIL import Image
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

In [11]:
# Define the number of subjects and conditions

num_subjects = 15
num_conditions = 6
num_images = num_subjects * num_conditions
img_width=320
img_height=243

creating dataset with Grayscale images of 15 subjects under six different conditions.

In [44]:
yale_data = np.zeros((img_width*img_height), dtype=np.float32)
for dirname, _, filenames in os.walk('/kaggle/input'):
    for fn in sorted(filenames):
        if fn[-3:] == 'png':
            image_filename = os.path.join(dirname, fn)
            print(image_filename)
            img = Image.open(image_filename)
            # Converting the image to grayscale and reshaping the array 
            img_arr = np.array(img).reshape(-1)
            yale_data=np.vstack([yale_data,img_arr])

/kaggle/input/removed-yale/yalefaces/s01cl.png
/kaggle/input/removed-yale/yalefaces/s01glass.png
/kaggle/input/removed-yale/yalefaces/s01happy.png
/kaggle/input/removed-yale/yalefaces/s01ll.png
/kaggle/input/removed-yale/yalefaces/s01ng.png
/kaggle/input/removed-yale/yalefaces/s01norm.png
/kaggle/input/removed-yale/yalefaces/s02cl.png
/kaggle/input/removed-yale/yalefaces/s02glass.png
/kaggle/input/removed-yale/yalefaces/s02happy.png
/kaggle/input/removed-yale/yalefaces/s02ll.png
/kaggle/input/removed-yale/yalefaces/s02ng.png
/kaggle/input/removed-yale/yalefaces/s02norm.png
/kaggle/input/removed-yale/yalefaces/s03cl.png
/kaggle/input/removed-yale/yalefaces/s03glass.png
/kaggle/input/removed-yale/yalefaces/s03happy.png
/kaggle/input/removed-yale/yalefaces/s03ll.png
/kaggle/input/removed-yale/yalefaces/s03ng.png
/kaggle/input/removed-yale/yalefaces/s03norm.png
/kaggle/input/removed-yale/yalefaces/s04cl.png
/kaggle/input/removed-yale/yalefaces/s04glass.png
/kaggle/input/removed-yale/yalefa

In [45]:
# checking the shape of yale faces data

yale_data=yale_data[1:][:]
print(yale_data)
print(yale_data.shape)

[[130. 130. 130. ...  68.  68.  68.]
 [130. 130. 130. ...  68.  68.  68.]
 [130. 130. 130. ...  68.  68.  68.]
 ...
 [108. 116. 117. ...  68.  68.  68.]
 [130. 130. 130. ...  68.  68.  68.]
 [130. 130. 130. ...  68.  68.  68.]]
(90, 77760)


So there are total 90 images comprising of 15 subjects or persons in which each subject has 6 conditions and the total number of pixels in each image is 77760.

As only one representative image can be stored for each subject in the database for future automated facial recognition purpose, PCA is applied to obtain the representative image for each person.i.e., Instead of six images for each person or subject we will extract one representative image for each person using PCA. So there will be total 15 representative images for 15 different persons with 77760 pixels for eachimage.

In [46]:
# Apply PCA to the yale_data matrix

pca_data=np.zeros((img_width*img_height), dtype=np.float32)

print("Explained variance ratios for 15 subjects are: ")

for i in range(0,90,6):
    pca = PCA(n_components=1)
    pca.fit(yale_data[i:i+6][:].T)
    
    print(pca.explained_variance_ratio_)

    rep_imgs = pca.transform(yale_data[i:i+6][:].T)
    rep_imgs=np.concatenate(rep_imgs)
    
    pca_data=np.vstack([pca_data,rep_imgs])
    
pca_data=pca_data[1:]

print("15 representative images for 15 different persons data matrix is: \n",pca_data)

print(pca_data.shape)

Explained variance ratios for 15 subjects are: 
[0.75118228]
[0.86461801]
[0.7843508]
[0.8777496]
[0.84251602]
[0.69865881]
[0.83255763]
[0.80030599]
[0.70669852]
[0.82939319]
[0.85801907]
[0.76381105]
[0.8645282]
[0.85610167]
[0.76558934]
15 representative images for 15 different persons data matrix is: 
 [[141.028    141.028    141.028    ... 292.17532  292.17532  292.17532 ]
 [157.13399  153.74744  132.12598  ... 273.0994   273.0994   273.0994  ]
 [127.04791  127.04791  127.04791  ... 278.4969   278.4969   278.4969  ]
 ...
 [139.60948  139.60948  139.60948  ... 291.44     291.44     291.44    ]
 [126.68713  126.68713  126.68713  ... 277.874    277.874    277.874   ]
 [125.04704  125.918076 131.94208  ... 266.1075   266.1075   266.1075  ]]
(15, 77760)


As the facial recognition method is based on the smallest Euclidean distance between the image and the representative images in the database, we will calculate the distance between the representative images and the respective subject images.

In [47]:
eucledian_distance=[]
k=0
for j in range(15):
    distance=[]
    for i in range(k,k+6):
        distance.append(np.sqrt(np.sum(np.square(pca_data[j] - yale_data[i][:]))))
    eucledian_distance.append(distance)
print(eucledian_distance)
print(len(eucledian_distance))

[[86467.4, 86549.414, 86568.16, 79927.73, 85794.695, 86596.48], [92653.34, 91451.89, 91400.445, 87376.04, 93840.41, 91483.234], [92274.75, 93271.9, 93342.85, 84978.46, 91565.8, 92724.96], [96535.375, 96163.13, 96118.87, 89817.03, 96314.3, 96234.65], [85145.1, 85860.93, 85690.4, 78930.19, 86490.766, 85789.87], [77268.9, 80829.0, 80788.766, 63645.95, 79392.336, 80586.414], [94775.4, 95998.88, 96253.695, 86910.234, 93849.37, 95580.445], [88892.695, 90713.055, 90779.586, 80730.61, 88386.09, 90203.32], [84678.94, 86498.48, 86611.38, 77103.73, 84888.54, 86154.95], [95661.0, 96680.51, 96745.5, 87931.31, 94369.38, 96537.97], [97253.64, 96616.516, 96659.72, 91180.516, 96758.83, 96465.59], [86653.91, 86228.12, 86009.21, 83074.51, 88587.5, 86239.21], [93504.12, 94262.46, 94230.414, 86164.695, 94160.28, 93758.16], [86580.85, 88748.1, 88951.51, 75936.016, 86909.67, 88299.695], [90379.586, 91618.92, 91687.2, 82428.92, 89354.37, 91177.41]]
15


From the above output, we can obtain the eucledian distances for all the 90 images in the 15 subjects under 6 conditions with their respective representative images.

To determine the number of images out of 90 that we are able to correctly identify based on this approach, we need to set a threshold value for the Euclidean distance. The values less than the threshold means that the representative image can be used for the Facial Recognition purpose.

In [53]:
threshold=95000
c=0
for i in eucledian_distance:
    for j in i:
        if j<=threshold:c=c+1
print(c)

73


So we can predict 73 images out of 90 correctly.

**CONCLUSION:**

The threshold for determining whether two images are similar or dissimilar based on their Euclidean distance depends on the specific application and the context of the analysis.

In some cases, a small Euclidean distance (e.g., less than a certain threshold value) may be used to indicate that two images are similar or nearly identical, while a large distance may indicate that the images are dissimilar.

However, in other cases, the threshold may be more subjective or dependent on the specific use case. For example, if the images being compared are of different resolutions or sizes, or if they have different lighting conditions or angles, then the threshold may need to be adjusted accordingly.

It is important to note that the Euclidean distance is just one of many possible methods for comparing the similarity or dissimilarity between images, and the choice of distance metric and threshold will depend on the specific needs and constraints of the analysis.